Fix ComplementaryArea's initial isActive state

WunderBart opened the pull request:
## What?
- Define the loading state for the `ComplementaryArea` component,
- Update related E2E utilities.

## Why?
The active complementary area identifier is initially unavailable as it needs to be loaded into the store. This leads to a situation where the initial `isActive` state of the `ComplementaryArea` component is `false` and, after one or more cycles, becomes `true`.

While this change doesn't really manifest itself visually (it lasts only a few milliseconds), it is a good practice in the E2E testing context because it reliably signals when the component is ready to interact.

For example, if an E2E test wants to open the `Editor settings` sidebar (which uses the `ComplementaryArea` component), it needs to check whether that sidebar is already opened before it clicks the toggle button. If that sidebar's initial `isActive` state is `false`, but the actual state is `true` (which will kick in with a delay), we might end up closing the sidebar instead if the click is late by a few milliseconds.

Also, as @kevin940726 [mentioned](https://github.com/WordPress/gutenberg/pull/47498#pullrequestreview-1274506872), this change "might also have a positive impact for assistive technologies, which they usually evaluate the context of the page as soon as possible for performance reasons."

## How to test

The E2E tests should cover this change, so it shouldn't need manual testing.

## Related
- https://github.com/WordPress/gutenberg/pull/43506
- https://github.com/WordPress/gutenberg/pull/46467#issuecomment-1404965411
- https://github.com/WordPress/gutenberg/pull/40923

WunderBart committed:
Reflect the actual initial state of the ComplementaryArea

WunderBart committed:
Update openDocumentSettingsSidebar to wait until button is enabled

Mamaduka commented on packages/interface/src/components/complementary-area/index.js:
```diff
@@ -144,6 +148,7 @@ function ComplementaryArea( {
 								isActive && ( ! showIconLabels || isLarge )
 							}
 							aria-expanded={ isActive }
+							disabled={ isActive === undefined }
```
I believe it's recommended to use `aria-disabled` for a11y so that disabled buttons can receive focus.

WunderBart replied:
https://github.com/WordPress/gutenberg/pull/47498/commits/0af516e80885a61b369501cb1e6029706099ce61

WunderBart committed:
Use `aria-disabled` instead of `disabled`

Mamaduka commented on packages/e2e-test-utils-playwright/src/editor/open-document-settings-sidebar.ts:
```diff
@@ -5,14 +5,18 @@ import type { Editor } from './index';
  *
  * @param {Editor} this
  */
 export async function openDocumentSettingsSidebar( this: Editor ) {
```
I think we have a similar helper in Puppeteer utils. Should we update it as well?

WunderBart replied:
Good point! Updated in https://github.com/WordPress/gutenberg/pull/47498/commits/725ad50f5c0bfcd46f646f150de0171374f623dd.

Mamaduka commented:
@WunderBart, the changes look good to me 👍  Do you have manual testing instructions, or should we rely on CI tests?

WunderBart committed:
Update also on Puppeteer's side

WunderBart commented:
> @WunderBart, the changes look good to me 👍 Do you have manual testing instructions, or should we rely on CI tests?

Thanks for the review, @Mamaduka! 🙇 I think CI should tell us (based on the tests using the Settings sidebar) if the change is OK. 🤞

jsnajdr commented on packages/e2e-test-utils-playwright/src/editor/open-document-settings-sidebar.ts:
```diff
@@ -5,14 +5,18 @@ import type { Editor } from './index';
+		.getByRole( 'button', {
+			name: 'Settings',
+			disabled: false,
+		} );
```
Is it preferred to write locators this way, with chained methods instead of the Playwright selector language, or whatever that `role=...` string is? (I couldn't find any docs on this syntax earlier today when I tried.)

WunderBart replied:
I'd say that it's the preferred way according to Playwright, as it's what is suggested when using their [built-in test generator](https://playwright.dev/docs/getting-started-vscode#generating-tests). It's more readable, flexible, and robust compared to text selectors. In this specific case, for example, the `disabled` value is automatically derived from either `disabled` or `aria-disabled` attribute, which would need to be explicitly requested via a classic CSS selector.

jsnajdr replied:
What confuses me, however, is that the `role=region[name="x"]` string is not a CSS or XPath selector, but something else, Playwright-specific. I didn't find docs for it, but it maps 1:1 to the `getRole( 'region', { name: 'x' } )` syntax.

WunderBart replied:
It's probably because they're not recommending those selectors anymore, so they've removed it from the documentation. 🤷 I can see a lot of `not recommended` stuff in the [Other selectors](https://playwright.dev/docs/other-locators) section now, which seems to have been updated with the [recent releases](https://playwright.dev/docs/release-notes) (we're currently using v1.27). I can also see that locator chaining is now [TheRecommendedWay™](https://playwright.dev/docs/locators#chaining-locators). That documentation changes very often, and I can't see a way of going back, at least a version or two, which has also been a pain point for me. 😕

kevin940726 replied:
Seems like they removed the doc for the text-based `role=` selector 😞 . It should be a one-to-one mapping to the now recommended `getByRole` though. We probably would want to migrate the existing usages when we upgrade playwright, and update our [doc](https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/e2e/README.md#use-accessible-selectors).

Mamaduka replied:
It's bad that they completely removed the docs for `page.locator`. The text-based locators are more flexible in some cases. Should we file an issue to restore them, at least on some level?

kevin940726 replied:
We can try that but I think they reached a decision already 😅 . I prefer text-based role selectors in most cases too, but I also don't mind using `getByRole` if that's what is officially recommended.

jsnajdr replied:
> We can try that but I think they reached a decision already

If they want to deprecate the text-based `role=` selectors, that's fine, no problem with that decision,  but removing the documentation is a strictly bad idea. Existing codebases still use them and they still work, so they should be documented.

jsnajdr commented on packages/interface/src/components/complementary-area/index.js:
```diff
@@ -144,6 +148,7 @@ function ComplementaryArea( {
 								isActive && ( ! showIconLabels || isLarge )
 							}
 							aria-expanded={ isActive }
```
The `aria-expanded` attribute (and possibly others) should always get a boolean value instead of `undefined`.

WunderBart replied:
`undefined` is listed as one of the allowed values for the `aria-expanded` property, as it defines an "[element that does not own or control an expandable grouping element](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-expanded#values)." This is also validated by the [jsx-a11y/aria-proptypes](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/93f78856655696a55309440593e0948c6fb96134/docs/rules/aria-proptypes.md) rule.

jsnajdr reviewed (commented):
This looks like a good fix :+1:

The main issue is that on editor boot, enabling the sidebar is a two-step process:
1. There is a server-supplied preference, `isComplementaryAreaVisible`. Server sends that in the initial HTML and when the sidebar is enabled, this value is `true` from the very beginning.
2. The sidebar has two tabs, with ids `edit-post/document` and `edit-post/block`. Initially none is selected, and the `activeArea` is `undefined`, although the sidebar is opened.
3. Only a bit later, in an effect, the `activeArea` is set to a valid value. Either in an effect that sets the `isActiveByDefault` value, or in an effect that observes block selection and selects the document ("Post", "Template", ...) tab when no block is selected, and switches to "Block" once a block gets selected.

This PR catches that ambiguous "sidebar opened but no tab selected" situation and disables the toggle button while it lasts.

WunderBart committed:
Make naming consistent

WunderBart committed:
Fix the initial `isActive` state in mobile viewport
The panel is always initially closed in mobile viewport

kevin940726 reviewed (approved):
I'm not sure but this fix might also have a positive impact for assistive technologies which they usually evaluate the context of the page as soon as possible for performance reasons. Anyway it LGTM! 💯

WunderBart committed:
Use `waitFor` instead of `expect`
This makes more sense contextually. `expect` should be reserved for the test subject.

WunderBart committed:
Rename "Close settings sidebar" to "Close settings"
This is consistent with the current settings sidebar name in the editor.

WunderBart committed:
Check if the area is being activated
For a short time, if the `isComplementaryAreaVisible` preference is set to true, the `activeArea` value is `undefined`. Let's disable the area toggle button until the area is activated properly. This will help with E2E stability because the framework will wait until the button is enabled, and can also help with a11y as it will correctly indicate the initial state of the toggle.

WunderBart committed:
Initially set the area disabled for small screens
This makes the area state values to be consistent.

WunderBart committed:
Reword comment

WunderBart committed:
Create isComplementaryAreaLoading selector

WunderBart commented on packages/edit-site/src/components/sidebar-edit-mode/index.js:
```diff
@@ -69,7 +69,7 @@ export function SidebarComplementaryAreaFills() {
 				identifier={ sidebarName }
 				title={ __( 'Settings' ) }
 				icon={ isRTL() ? drawerLeft : drawerRight }
-				closeLabel={ __( 'Close settings sidebar' ) }
```
It's called `Close settings` everywhere else, so it made sense to me to make it consistent.

WunderBart commented:
👋 @kevin940726 @jsnajdr @talldan, I'd appreciate it if you could take another look - I needed to make some updates to address the persisting area visibility preference between viewports.

WunderBart commented:
👋 @kevin940726 friendly ping in case you missed the last one! Could you take another look at this one? 🙏

kevin940726 reviewed (approved):
Change looks good to me 👍 Thanks!
